Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XlgDebugger/JPath/JPath.g4

```diff
@@ -0,0 +1,149 @@
+NMATCH  : '!~' ;
+
+// arithmetic operators
+MINUS   : '-'  ;
+PLUS    : '+'  ;
+TIMES   : '*'  ;
+DIV     : '/'  ;
+MOD     : '%'  ;
+
+// array operators
+CONCAT    : '++' ;
+INTERSECT : '&'  ;
+
+IntLit
+    : [0-9]+ ;
+
+StrLit
+    : '\'' ~[']* '\''
+    | '"' ~["]* '"'
+    ;
+
+RegExLit
+    : '/' ~[/]+ '/' 
+    | '!' ~[!]+ '!'
+    ;
+
+fragment IdFragment
+    : [a-zA-Z][a-zA-Z0-9_]* ;
+
+PropertyId
+    : IdFragment ;
+
+VarId
+    : '$' IdFragment ;
+
+EscID
+    : '`' ~[`]+ '`' ;
+
+Opt
+    : '-' [a-zA-Z0-9'-']+ ;
+
+intBinaryOp
+    : Token=(PLUS | MINUS | TIMES | DIV | MOD) ;
+
+intUnaryOp
+    : Token=MINUS ;
+
+boolBinaryOp
+    : Token=(GTE | GT | LTE | LT | EQ | NEQ | MATCH | NMATCH) ;
+    
+logicBinaryOp
+    : Token=(AND | OR | XOR | IFF) ; 
+
+logicUnaryOp
+    : Token=NOT ;
+
+arrayBinaryOp
+    : Token=(CONCAT | INTERSECT) ;
+
+anyBinaryOp
+    : intBinaryOp
+    | boolBinaryOp
+    | logicBinaryOp
+    | arrayBinaryOp
+    ;
+
+intExpr
+    : Expr=expr                                       #ExprIntExpr
+    | Op=intUnaryOp Sub=intExpr                       #UnaryIntExpr
+    | Lhs=intExpr Op=intBinaryOp Rhs=intExpr          #BinaryIntExpr
+    | '(' Sub=intExpr ')'                             #SubIntExpr
+    ;
+
+boolExpr
+    : Lhs=intExpr Op=boolBinaryOp Rhs=intExpr         #BinaryBoolExpr
+    | '(' Sub=boolExpr ')'                            #SubBoolExpr
+    ;
+
+logicExpr
+    : Expr=boolExpr                                   #BoolLogicExpr
+    | Lhs=logicExpr Op=logicBinaryOp Rhs=logicExpr    #BinaryLogicExpr
+    | Op=logicUnaryOp Sub=logicExpr                   #UnaryLogicExpr
+    | '(' Sub=logicExpr ')'                           #SubLogicExpr
+    ;
+
+prop
+    : PropertyName=PropertyId                         #PropertyId
+    | PropertyName=EscID                              #EscId
+    ;
+
+selector
+    : Name=prop                                       #IdSelector
+    | '(' Names+=prop ('+' Names+=prop)+ ')'          #UnionSelector
+    ;
+
+literal
+    : Value=StrLit                                    #StrLitExpr
+    | Value=RegExLit                                  #RegExLitExpr
+    | Value=IntLit                                    #IntLitExpr
+    ;
+
+expr
+    : '$'                                             #RootExpr
+    | Var=VarId                                       #VarExpr
+    | Sub=selector                                    #SelectorExpr
+    | Lit=literal                                     #LiteralExpr
+    | Lhs=expr '.' Selector=selector                  #MapExpr
+    | Lhs=expr '[' Filter=logicExpr ']'               #FilterExpr
+    | Lhs=expr '[' Index=intExpr ']'                  #IndexExpr
+    | Lhs=expr '[' Begin=intExpr '..' End=intExpr ']' #RangeExpr
+    | '#' Sub=expr                                    #CardinalityExpr
+    | Func=expr '(' Args+=expr (',' Args+=expr)* ')'  #FuncAppExprParen
+    | Func=expr Arg=expr                              #FuncAppExpr
```

Comment:
Having space as function application seems very fragile and hard to learn and read. I still get confused about powershells sometimes parens, sometimes space argument separation. Should we just keep it consistent and straight-forward with just parens?

Reply:
sounds good, will change